### PR TITLE
Refactor public builder API

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves https://github.com/jaegertracing/jaeger-client-java/issues/258 #44 
Supersedes https://github.com/jaegertracing/jaeger-client-java/pull/336

The idea is to use `Configuration` only for creating `Tracer.Builder` from environmental variables (or in the future from whatever source).

Reporter, Sender, Sampler classes are arbitrary difficult to construct therefore they _might_ implement a builder interface defined in main Reporter, Sender, Sampler interface class.

Other high-level concepts:
* tracer builder is opinionated and without any configuration it creates a tracer with service name `unknown` and sends data to the local agent.
* impl classes like senders, reporter holds the default configuration e.g. default flush interval, default hostname of the agent.

This is WIP without javadoc, tests. Please just look at the overall approach whether it is something we want.

### CODE REVIEW

**Comment by pavolloffay** [comment]
cc @jpkrohling @yurishkuro

File: jaeger-core/src/main/java/com/uber/jaeger/Tracer.java [hunk 1]
```diff
@@ -449,6 +453,10 @@ public Scope startActive(boolean finishSpanOnClose) {
     private ScopeManager scopeManager = new ThreadLocalScopeManager();
     private BaggageRestrictionManager baggageRestrictionManager = new DefaultBaggageRestrictionManager();
 
+    public Builder() {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Tracer.java]: maybe `Tracer.builder(serviceName)` instead? What are the conventions in Java these days?  I personally prefer static methods to constructors.

File: jaeger-core/src/main/java/com/uber/jaeger/Tracer.java [hunk 2]
```diff
@@ -529,7 +547,22 @@ public Builder withBaggageRestrictionManager(BaggageRestrictionManager baggageRe
     }
 
     public Tracer build() {
-      return new Tracer(this.serviceName, reporter, sampler, registry, clock, metrics, tags,
+      if (serviceName == null) {
+        serviceName = "unknown";
+        log.warn("Service name hasn't been configured, using " + serviceName + " instead");
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Tracer.java]: What is the argument for allowing "unknown"? Service name is a fundamental concept in Jaeger, I would rather require it in the `builder()` func.

File: jaeger-core/src/main/java/com/uber/jaeger/reporters/RemoteReporter.java [hunk 3]
```diff
@@ -187,4 +189,56 @@ public void close() {
...
+    public Reporter build() {
+      if (sender == null) {
+        sender = new UdpSender(UdpSender.DEFAULT_AGENT_UDP_HOST, UdpSender.DEFAULT_AGENT_UDP_COMPACT_PORT, 0);
+      }
+      if (metrics == null) {
+        metrics = new Metrics(new StatsFactoryImpl(new NullStatsReporter()));
+      }
+      Reporter reporter = new RemoteReporter(sender, flushInterval, maxQueueSize, metrics);
+      if (withLogging) {
+        reporter = new CompositeReporter(new LoggingReporter(), reporter);
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/reporters/RemoteReporter.java]: This is one reason why the separation of configuration exists. I think it breaks encapsulation if RemoteReporter suddenly becomes aware of things like logging reporter.

File: jaeger-core/src/main/java/com/uber/jaeger/samplers/Sampler.java [hunk 4]
```diff
@@ -29,4 +29,8 @@
    * Release any resources used by the sampler.
    */
   void close();
+
+  interface Builder {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/samplers/Sampler.java]: why is this needed?

File: jaeger-core/src/main/java/com/uber/jaeger/senders/UdpSender.java [hunk 5]
```diff
@@ -32,6 +32,13 @@
   private Agent.Client agentClient;
   private ThriftUdpTransport udpTransport;
 
+  /**
+   * This constructor expects Jaeger running running on {@value #DEFAULT_AGENT_UDP_HOST} and port {@value #DEFAULT_AGENT_UDP_COMPACT_PORT}
+   */
+  public UdpSender() {
+    this(DEFAULT_AGENT_UDP_HOST, DEFAULT_AGENT_UDP_COMPACT_PORT, 0);
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/UdpSender.java]: is `maxPacketSize=0` valid?

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java [hunk 6]
```diff
@@ -175,10 +175,12 @@
    */
   private Tracer tracer;
 
+  @Deprecated
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: does this still allow populating Configuration object from a config file?

File: jaeger-core/src/main/java/com/uber/jaeger/senders/UdpSender.java (hunk 5 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/senders/UdpSender.java]: It is, look how it was called from the configuration class. I don't like that default values are defined in configuration classes.

File: jaeger-core/src/main/java/com/uber/jaeger/samplers/Sampler.java (hunk 4 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/samplers/Sampler.java]: The reason is mentioned in PR description. Some samplers are easy to create, however some (like `RemoteControlledSampler`) has several configuration options therefore there is a generic builder.

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java (hunk 6 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: Ignore this, I want to say that only `Tracer.Builder builder = Configuration.fomEnv()` should be used. See PR description

File: jaeger-core/src/main/java/com/uber/jaeger/reporters/RemoteReporter.java (hunk 3 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/reporters/RemoteReporter.java]: It can be removed from here, I think it's better to add it directly in `TracerBuilder`

File: jaeger-core/src/main/java/com/uber/jaeger/Tracer.java (hunk 2 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/Tracer.java]: +1 there si a comment in a PR description about this

**Comment by pavolloffay** [comment]
@yurishkuro thanks for the review. Do you think we want to proceed with this approach? Don't do nit pitching, just see the overall approach described in the first comment in this PR.

File: jaeger-core/src/main/java/com/uber/jaeger/samplers/Sampler.java (hunk 4 above)
Comment (yurishkuro) [reply jaeger-core/src/main/java/com/uber/jaeger/samplers/Sampler.java]: when someone creates a concrete sampler/reporter, they are dealing with a concrete class. If it happened to have a builder, it would be specific to that class. So I am not seeing the value of a common builder interface, unless the intention is to pass builders around.

**Comment by yurishkuro** [comment]
I do like the idea of coalescing defaults into one place. Whether that's configs or builders - no strong opinion.

The original issue was about config classes being difficult to construct programmatically. I think this PR's approach simply bypasses that problem by offering construction via builders directly in the components, i.e. if you're constructing programmatically you don't need to use the configs. +1 to that.

My only additional requirement is that changes to configs should be kept backwards compatible (deprecated - fine). The primary use case for config class at Uber is when it is populated from a config file, so that should still work (not sure if we have tests for that, maybe in dw module).

**Comment by pavolloffay** [comment]
I won't remove config classes in this PR. internals in config classes will use these new builders. I will iterate over tomorrow to the final state.

The high level idea of this PR is to move builders to impl classes so impl classes are responsible of creating. Then there is a configuration layer which just maps configfiles/env/props to these builders.

> The primary use case for config class at Uber is when it is populated from a config file, so that should still work (not sure if we have tests for that, maybe in dw module). 

Are is it possible to share this? e.g. if there is a code which loads yaml/json and creates Jaeger tracer that could be moved here I guess.

**Comment by pavolloffay** [comment]
I have moved all default values to impl classes.

To summarize this:
* default values are moved to impl classes
* Configuration classes are dedicated to hold and set/get only primitive values and in the end create tracer.
* Configuration supports creating itself (then the tracer) from env vars or `Map<string, string>` (to be done separately)
* Configuration supports creating from env and then overriding any property 
* Complicated classes with more properties have builders

Here is a code snippet:
```java
    // Tracer Builder directly
    Tracer tracer = new Tracer.Builder("name")
        .withReporter(new Builder()
            .withSender(new HttpSender("http://localhost:8080"))
            .withFlushInterval(100)
            .build())
        .withStatsReporter(new NullStatsReporter())
        .withSampler(new RemoteControlledSampler.Builder("name")
            .withInitialSampler(new ConstSampler(true))
            .build())
        .build();

    // Configuration from env
    System.setProperty(Configuration.JAEGER_SERVICE_NAME, "name");
    Configuration configuration = Configuration.fromEnv();
    configuration.getReporterConfig()
        .getSenderConfiguration()
        .getBuilder()
        .endpoint("http://jaegereverywhere.com");
    configuration.getTracer();

    // Configuration directly
    new Configuration("name")
        .reporterConfiguration(new ReporterConfiguration()
            .senderConfiguration(new SenderConfiguration.Builder()
                .endpoint("http://jaegereverywhere.com")
                .build()))
      .getTracer();
  }
```

**Comment by pavolloffay** [comment]
@yurishkuro @vprithvi @jpkrohling @objectiser could you please have a look?

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java [hunk 7]
```diff
@@ -263,56 +261,79 @@ public void setStatsFactory(StatsFactory statsFactory) {
     this.statsFactory = statsFactory;
   }
 
+  public Configuration reporterConfiguration(ReporterConfiguration reporterConfig) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: nit: naming not consistent with setStatsFactory or with builders. Any reason not to converge on `withXyz` naming?

nit2: Configuration suffix can probably be dropped

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java [hunk 8]
```diff
@@ -633,28 +641,6 @@ public static SenderConfiguration fromEnv() {
...
-                  chain.request()
-                          .newBuilder()
-                          .addHeader("Authorization", headerValue)
-                          .build()
-          );
-        }
-      };
-    }
-
     public static class Builder {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: can we deprecate this class and just use SenderConfiguration directly?

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java [hunk 9]
```diff
@@ -574,39 +592,29 @@ private SenderConfiguration(SenderConfiguration.Builder builder) {
...
-        return new HttpSender(endpoint, builder.build());
+        return httpSenderBuilder.build();
       }
 
       log.debug("Using the UDP Sender to send spans to the agent.");
       return new UdpSender(
-              stringOrDefault(this.agentHost, UdpSender.DEFAULT_AGENT_UDP_HOST),
-              numberOrDefault(this.agentPort, UdpSender.DEFAULT_AGENT_UDP_COMPACT_PORT).intValue(),
+              stringOrDefault(this.builder.agentHost, UdpSender.DEFAULT_AGENT_UDP_HOST),
+              numberOrDefault(this.builder.agentPort, UdpSender.DEFAULT_AGENT_UDP_COMPACT_PORT).intValue(),
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: should the defaults be moved into respective builders, like UdpSender.Builder?

File: jaeger-core/src/main/java/com/uber/jaeger/Tracer.java [hunk 10]
```diff
@@ -438,25 +442,21 @@ public Scope startActive(boolean finishSpanOnClose) {
...
-    public Builder(String serviceName, Reporter reporter, Sampler sampler) {
-      if (serviceName == null || serviceName.trim().length() == 0) {
-        throw new IllegalArgumentException("serviceName must not be null or empty");
-      }
-      this.serviceName = serviceName;
-      this.reporter = reporter;
-      this.sampler = sampler;
-
+    public Builder(String serviceName) {
+      this.withServiceName(serviceName);
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Tracer.java]: Q: what is the scenario where withServiceName() method is needed instead of providing the name to the ctor?

File: jaeger-core/src/main/java/com/uber/jaeger/Tracer.java [hunk 11]
```diff
@@ -529,9 +566,29 @@ public Builder withBaggageRestrictionManager(BaggageRestrictionManager baggageRe
...
+        reporter = new RemoteReporter.Builder()
+            .withMetrics(metrics)
+            .build();
+      }
+      if (sampler == null) {
+        sampler = new RemoteControlledSampler.Builder(serviceName)
+            .withMetrics(metrics)
+            .build();
+      }
+      if (withLoggingSpans) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Tracer.java]: copying from other PR: I prefer not to have this option on the Tracer.Builder, it creates unnecessary coupling with CompositeReporter, LoggingReporter, etc. This feature can be left to Config classes only.

File: jaeger-core/src/main/java/com/uber/jaeger/reporters/Reporter.java [hunk 12]
```diff
@@ -24,4 +24,8 @@
   void report(Span span);
 
   void close();
+
+  interface Builder<T extends Reporter> {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/reporters/Reporter.java]: I do not believe this interface is needed. No class actually uses it.

File: jaeger-core/src/main/java/com/uber/jaeger/samplers/Sampler.java [hunk 13]
```diff
@@ -29,4 +29,8 @@
    * Release any resources used by the sampler.
    */
   void close();
+
+  interface Builder<T extends Sampler> {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/samplers/Sampler.java]: similarly, this is never used on the left hand side.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/Sender.java [hunk 14]
```diff
@@ -23,4 +23,8 @@
   int flush() throws SenderException;
 
   int close() throws SenderException;
+
+  interface Builder<T extends Sender> {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/Sender.java]: never used

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java (hunk 8 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: I am definitely +1 on that, I will try to do it

File: jaeger-core/src/main/java/com/uber/jaeger/Tracer.java (hunk 11 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/Tracer.java]: I would recommend to people start using `Tracer.Builder` When they use programmatic API. I think it's useful to have it here directly. I am not sure if the composite reporter is easy enough to find/use.

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java [hunk 15]
```diff
@@ -490,15 +553,15 @@ public String getAgentHost() {
         return null;
       }
 
-      return this.senderConfiguration.agentHost;
+      return this.senderConfiguration.builder.agentHost;
```
Comment (objectiser) [inline jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: Possibly we should deprecate these accessors (getAgentHost/getAgentPort) as looks like they are only used in tests?

File: jaeger-core/src/main/java/com/uber/jaeger/reporters/RemoteReporter.java [hunk 16]
```diff
@@ -187,4 +194,42 @@ public void close() {
...
+
+    public Builder withSender(Sender sender) {
+      this.sender = sender;
+      return this;
+    }
+
+    @Override
+    public RemoteReporter build() {
+      if (sender == null) {
+        sender = new UdpSender(UdpSender.DEFAULT_AGENT_UDP_HOST, UdpSender.DEFAULT_AGENT_UDP_COMPACT_PORT, 0);
```
Comment (objectiser) [inline jaeger-core/src/main/java/com/uber/jaeger/reporters/RemoteReporter.java]: Should just use the default constructor.

File: jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java [hunk 17]
```diff
@@ -110,4 +122,53 @@ public void send(Process process, List<Span> spans) throws TException {
...
+      this.authInterceptor = getAuthInterceptor("Bearer " + authToken);
+      return this;
+    }
+
+    @Override
+    public HttpSender build() {
+      OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder();
+      if (authInterceptor != null) {
+        clientBuilder.addInterceptor(authInterceptor);
+
```
Comment (objectiser) [inline jaeger-core/src/main/java/com/uber/jaeger/senders/HttpSender.java]: nit: remove line

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java (hunk 15 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: I was thinking about this, We definitely need accessors like `Configuration.getReporterConfiguration` then `getSenderConfiguration` to be able to override sender without re-creating other configs.

I am wondering whether there is an use case when app gets the configuration from env and then changes behaviour - e.g. if used endpoint it will get it and override to https for example.

**Comment by pavolloffay** [comment]
> general comment, just my preference but what if we make all new builder constructors private and instead have a static builder() method on the respective class, e.g. Tracer.builder(), RepoteReporter.builder()? I think it improves readability of initialization code.

Can do, we can talk about this cosmetics at the end.

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java (hunk 15 above)
Comment (objectiser) [reply jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: Don't think we should be supporting such use cases - but if someone wanted to make a case for it later, it could be discussed then.

**Comment by pavolloffay** [comment]
I was able to rename conf method to `with` variants and deprecate `SenderConfiguration.Builder`. Overall I think this is getting to be ready soon

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java [hunk 18]
```diff
@@ -252,67 +253,99 @@ public synchronized void closeTracer() {
...
 
+  /**
+   * @deprecated Use {@link #withStatsFactory(StatsFactory)} instead
+   */
+  @Deprecated
   public void setStatsFactory(StatsFactory statsFactory) {
     this.statsFactory = statsFactory;
   }
 
+  public Configuration withStatsFactory(StatsFactory statsFactory) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: an overlap with #349, I suggest we don't touch setStatsFactory in this PR and let it be deprecated in #349

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java [hunk 19]
```diff
@@ -508,13 +591,16 @@ public Integer getFlushIntervalMs() {
...
+    public SenderConfiguration getSenderConfiguration() {
+      return senderConfiguration;
+    }
   }
 
   /**
    * Holds the configuration related to the sender. A sender can be a {@link HttpSender} or {@link UdpSender}
    *
    */
-  @Getter
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: looks like a breaking change

Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: Added back, although I think this was added only for tests by @jpkrohling. I don't think we had a requirement of exposing it.

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java (hunk 18 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: reerted

**Comment by pavolloffay** [comment]
@yurishkuro yes, that is what I am trying. I will merge on green

File: jaeger-core/src/main/java/com/uber/jaeger/Configuration.java [hunk 20]
```diff
@@ -263,56 +269,92 @@ public void setStatsFactory(StatsFactory statsFactory) {
...
+    return this;
+  }
+
+  public Configuration withCodec(CodecConfiguration codecConfig) {
+    this.codecConfig = codecConfig;
+    return this;
+  }
+
+  public Configuration withTracerTags(Map<String, String> tracerTags) {
+    this.tracerTags = tracerTags;
```
Comment (jpkrohling) [inline jaeger-core/src/main/java/com/uber/jaeger/Configuration.java]: You should use a new map, like `new HashMap(tracerTags)`, so that the caller won't change *your* copy of the map.
